src/main/java/liquibase/ext/spanner/SpannerInsertOrUpdateGenerator.java

```diff
@@ -0,0 +1,99 @@
+/**
+ * Copyright 2020 Google LLC
+ *
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package liquibase.ext.spanner;
+
+import java.util.ArrayList;
+import liquibase.database.Database;
+import liquibase.exception.LiquibaseException;
+import liquibase.sql.Sql;
+import liquibase.sql.UnparsedSql;
+import liquibase.sqlgenerator.SqlGeneratorChain;
+import liquibase.sqlgenerator.core.InsertOrUpdateGenerator;
+import liquibase.statement.core.InsertOrUpdateStatement;
+
+public class SpannerInsertOrUpdateGenerator extends InsertOrUpdateGenerator {
+  @Override
+  public boolean supports(InsertOrUpdateStatement statement, Database database) {
+    return database instanceof CloudSpanner;
+  }
+
+  @Override
+  public Sql[] generateSql(InsertOrUpdateStatement insertOrUpdateStatement, Database database,
+      SqlGeneratorChain sqlGeneratorChain) {
+    // Cloud Spanner does not have an UPSERT / MERGE DML statement, so we will generate both an
+    // INSERT and an UPDATE statement. The INSERT statement will check whether the record already
+    // exists, and only insert it in that case. The record is always updated.
+    //
+    // The most efficient way to do this in Cloud Spanner would be to use an InsertOrUpdate
+    // mutation. That is however not supported through Liquibase, as it requires the update to use
+    // SQL.
+    ArrayList<Sql> sqlList = new ArrayList<>(2);
+    if (!insertOrUpdateStatement.getOnlyUpdate()) {
+      sqlList.add(
+          new UnparsedSql(getInsertStatement(insertOrUpdateStatement, database, sqlGeneratorChain),
+              "", getAffectedTable(insertOrUpdateStatement)));
+    }
+
+    String whereClause = getWhereClause(insertOrUpdateStatement, database);
+    try {
+      String update =
+          getUpdateStatement(insertOrUpdateStatement, database, whereClause, sqlGeneratorChain);
+      if (update.endsWith("\n") && update.length() > 1) {
```

Comment:
Some of this code I assume is taken from similar functions in core Liquibase? This is yet another way to remove a character from an end of a string.

Reply:
Yes. I tried to reuse the original `InsertOrUpdateGenerator`, but the (internal) API of it changed a little bit between version 3.x and 4.x, so it's not doable if we want to be able to support both. So I ended up copying the main parts of the SQL generation into this generator.